### PR TITLE
Bluetooth: Controller: Fix LLCP event_counter value used during prepare

### PR DESCRIPTION
**Pull request by cvinayak**
Fix LLCP to use event_counter value used during prepare to compare instant to be the value as when in prepare and not incorrectly use a stale value (when two radio events overlap).

### CODE REVIEW

File: subsys/bluetooth/controller/ll_sw/ull_conn.c [hunk 1]
```diff
@@ -2177,6 +2177,11 @@ void ull_conn_resume_rx_data(struct ll_conn *conn)
 }
 #endif /* CONFIG_BT_CTLR_LE_ENC */
 
+uint16_t ull_conn_event_counter_at_prepare(struct ll_conn *conn)
```
Comment (Thalley) [inline subsys/bluetooth/controller/ll_sw/ull_conn.c]:
```suggestion
uint16_t ull_conn_event_counter_at_prepare(const struct ll_conn *conn)
```

**Review by Thalley — DISMISSED** [review]
Minor suggestion, but otherwise LGTM

**Review by Thalley — DISMISSED** [review]

**Review by thoh-ot — CHANGES_REQUESTED** [review]
Need time to look at this.

**Comment by cvinayak** [comment]
> Need time to look at this.

The issue was discovered running the `central_gatt_write` and `peripheral_gatt_write` samples with conn update procedures that was added in #92572. The sample can be run in bsim to reproduce supervision timeout due to incorrect instant being checked when consecutive connection events (prepare requesting preemption) overlap under high throughput scenario.

**Comment by thoh-ot** [comment]
This PR removes fixes that were made as part of https://github.com/zephyrproject-rtos/zephyr/pull/52822 to stabilize CIS therefore I'm not really fond of it.

`ull_conn_event_counter_at_prepare()` is just the prepare path (TX path) of `ull_conn_event_counter()`, so something must be wrong with using `ull_ref_get() ` for the post-prepare path (RX path) of `ull_conn_event_counter()`.

I'm also a bit curious to why the LLCP unit-tests are passing with this change, either they aren't being run or they don't catch this

**Comment by cvinayak** [comment]
> This PR removes fixes that were made as part of #52822 to stabilize CIS therefore I'm not really fond of it.

```
/* Check if this connection event is where we need to start the CIS */
```
Above is the added fixes/improvement comment, but which can execute anyway inside a connection event anywhere upto 150 us before the next connection event. During the overlap of preemption timeout duration, the `ull_ref_get()` value used in `rp_cc_check_instant()` call done insde the event (not a prepare) is in that corner case.

> I'm also a bit curious to why the LLCP unit-tests are passing with this change, either they aren't being run or they don't catch this.

I do not think there is a test to cover the fixes/improvements added in #52822 where the call to `rp_cc_check_instant()` before next event and after prepare unto preempt is tested. The change in #52822 seems like a workaround compared to other control procedures that check instant in prepare and not during PDU reception.

**Comment by thoh-ot** [comment]
> > This PR removes fixes that were made as part of #52822 to stabilize CIS therefore I'm not really fond of it.
> 
> ```
> /* Check if this connection event is where we need to start the CIS */
> ```
> 
> Above is the added fixes/improvement comment, but which can execute anyway inside a connection event anywhere upto 150 us before the next connection event. During the overlap of preemption timeout duration, the `ull_ref_get()` value used in `rp_cc_check_instant()` call done insde the event (not a prepare) is in that corner case.
> 
> > I'm also a bit curious to why the LLCP unit-tests are passing with this change, either they aren't being run or they don't catch this.
> 
> I do not think there is a test to cover the fixes/improvements added in #52822 where the call to `rp_cc_check_instant()` before next event and after prepare unto preempt is tested. The change in #52822 seems like a workaround compared to other control procedures that check instant in prepare and not during PDU reception.

The original commit has a bit more explaining: https://github.com/zephyrproject-rtos/zephyr/pull/52822/commits/4f5b711d2381fd580aa07915a2ad5c42bda1d9ed

It's to handle the case where the CIS begins in the next event.

**Comment by cvinayak** [comment]
> It's to handle the case where the CIS begins in the next event.

Correction, i.e it is an improvement to handle cases where CIS begins in the current ACL event (and before next ACL event) where the CIS_IND was received. Without the improvement, the CIS will still be established in the next ACL event after calculating any instant latency and appropriately compensating for elapsed CIS intervals and CIS offsets (hence no CI tests failed with/without this improvement)

I have added back the improvement with renamed function names to bring clarity to when the instant is checked, state machine run at prepare versus on reception of CIS_IND PDU.

**Review by Thalley — COMMENTED** [review]
Is there any way to test this fix? CI passes both before and after

**Review by thoh-ot — APPROVED** [review]
I'm wondering if the usefulness of `ull_conn_event_counter()` is gone, now we know that the `ull_ref_get()` logic is flawed...

None the less, I'll approve these changes.

**Comment by cvinayak** [comment]
> I'm wondering if the usefulness of `ull_conn_event_counter()` is gone, now we know that the `ull_ref_get()` logic is flawed...
> 
> None the less, I'll approve these changes.

`ull_ref_get()` was intended to be used to count references when same state/role events overlap, like ACL connections occupying entire event length, directed advertising and continuous scanning so that state/role can be gracefully disabled when reference count reaches zero; definitely not to identify being inside an event, though it works. Yes, `ull_conn_event_counter()` is not useful in providing the event_counter, it is better to use the right event_counter value inlined based on where in the timeline the event_counter is needed.

That said, current use of `ull_conn_event_counter()` to assign instant value in initiating PDUs being in Central role is acceptable for now.

**Comment by cvinayak** [comment]
> Is there any way to test this fix? CI passes both before and after

Testable in unit test (otherwise flakky to test with a modified tester Controller implementation under bsim).

This PR is not fixing anything fatal, I can get back to adding unit test after some quiet period!

**Review by Thalley — APPROVED** [review]
